src/fast_check/transform_dts.rs

```diff
@@ -0,0 +1,1352 @@
+      TsFnOrConstructorType, TsFnParam, TsFnType, TsKeywordType,
+      TsKeywordTypeKind, TsPropertySignature, TsTupleElement, TsTupleType,
+      TsType, TsTypeAnn, TsTypeElement, TsTypeLit, VarDecl, VarDeclKind,
+      VarDeclarator,
+    },
+    common::DUMMY_SP,
+  },
+  ModuleSpecifier, ParsedSource, SourceRange, SourceRangedForSpanned,
+};
+
+use crate::{FastCheckDiagnostic, FastCheckDiagnosticRange};
+
+use super::swc_helpers::{
+  any_type_ann, maybe_lit_to_ts_type, maybe_lit_to_ts_type_const, ts_readonly,
+  ts_tuple_element, type_ann,
+};
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum FastCheckDtsDiagnostic {
+  #[error("unable to infer type from expression or declaration")]
+  UnableToInferType { range: FastCheckDiagnosticRange },
+  #[error("unable to infer type, falling back to any type")]
+  UnableToInferTypeFallbackAny { range: FastCheckDiagnosticRange },
+  #[error("unable to infer type from object property, skipping")]
+  UnableToInferTypeFromProp { range: FastCheckDiagnosticRange },
+  #[error("unable to infer type from spread, skipping")]
+  UnableToInferTypeFromSpread { range: FastCheckDiagnosticRange },
+  #[error("cannot infer type from using, skipping")]
+  UnsupportedUsing { range: FastCheckDiagnosticRange },
+}
+
+impl FastCheckDtsDiagnostic {
+  pub fn specifier(&self) -> &ModuleSpecifier {
+    match self {
+      FastCheckDtsDiagnostic::UnableToInferType { range } => &range.specifier,
+      FastCheckDtsDiagnostic::UnableToInferTypeFallbackAny { range } => {
+        &range.specifier
+      }
+      FastCheckDtsDiagnostic::UnableToInferTypeFromProp { range } => {
+        &range.specifier
+      }
+      FastCheckDtsDiagnostic::UnableToInferTypeFromSpread { range } => {
+        &range.specifier
+      }
+      FastCheckDtsDiagnostic::UnsupportedUsing { range } => &range.specifier,
+    }
+  }
+
+  pub fn range(&self) -> Option<&FastCheckDiagnosticRange> {
+    match self {
+      FastCheckDtsDiagnostic::UnableToInferType { range } => Some(range),
+      FastCheckDtsDiagnostic::UnableToInferTypeFallbackAny { range } => {
+        Some(range)
+      }
+      FastCheckDtsDiagnostic::UnableToInferTypeFromProp { range } => {
+        Some(range)
+      }
+      FastCheckDtsDiagnostic::UnableToInferTypeFromSpread { range } => {
+        Some(range)
+      }
+      FastCheckDtsDiagnostic::UnsupportedUsing { range } => Some(range),
+    }
+  }
+}
+
+pub struct FastCheckDtsTransformer<'a> {
+  id_counter: usize,
+  parsed_source: &'a ParsedSource,
+  pub diagnostics: Vec<FastCheckDtsDiagnostic>,
+  specifier: &'a ModuleSpecifier,
+}
+
+impl<'a> FastCheckDtsTransformer<'a> {
+  pub fn new(
+    parsed_source: &'a ParsedSource,
+    specifier: &'a ModuleSpecifier,
+  ) -> Self {
+    Self {
+      id_counter: 0,
+      parsed_source,
+      specifier,
+      diagnostics: vec![],
+    }
+  }
+
+  fn gen_unique_name(&mut self) -> String {
+    self.id_counter += 1;
+    format!("_dts_{}", self.id_counter)
+  }
```

Comment:
In the future we can make this better to actually avoid conflicts, but the chance of that happening are so slim (ex. do a pre-pass on the file and collect all the identifier names--including globals referenced in types and put them in a hashset, then here it would try something similar, but check for conflicts)